BYOP and DB2OnCloudSetup has out of date instructions for datasource configuration #2303 (#2449)
IBM/FHIR: docs/src/pages/guides/DB2OnCloudSetup.md

      "query_options": {
        "authSource": "admin",
        "replicaSet": "replset"
      },
      "replica_set": "replset",
      "scheme": "db2",
      "type": "uri"
    }
  },
  "iam_apikey_description": "Auto-generated for key username1-fc89-48e1-a1ce-569ca6c6e3d5",
  "iam_apikey_name": "Service credentials-1",
  "iam_role_crn": "crn:v1:bluemix:public:iam::::serviceRole:Manager",
  "iam_serviceid_crn": "secret",
  "instance_administration_api": {
    "deployment_id": "crn:v1:bluemix:public:dashdb-for-transactions:us-east:a/secret:",
    "instance_id": "crn:v1:bluemix:public:dashdb-for-transactions:us-east:a/secret::",
    "root": "https://api.us-east.db2.cloud.ibm.com/v4/ibm"
  }
}
```

4. Save these details for later, as these properties are needed to deploy and manage the IBM FHIR Server schema.

For improved security, the **admin** user should only be used to deploy the schema objects (tables, indexes, stored procedures etc) and administer the database, NOT for connecting from the IBM FHIR Server application.

### **Adding the FHIRSERVER user**

Following the least-privilege principle, the IBM FHIR server itself recommends running as db user, not db administrator.

> # Using Db2 Auth

Reviewer: I wonder if we should just recommend one or the other.  Maybe always go with the IAM approach?

Author: I think offering both makes sense to me. Since I seem to be troubleshooting folks using both.

> # Using Db2 Auth

Reviewer: should this be an h4?
```suggestion
#### Using Db2 Auth
```

1. On the Db2 Resource, click `Manage`

2. Click `Open Console`

3. Click Administration > User Management

4. Click Add User, and enter the user details with a sufficiently complex password. Set the `User Privilege` as `User`. 

5. Save these details for the datasource.xml.

> # Using IAM apiKey

Reviewer: should this be an h4?
```suggestion
#### Using IAM apiKey
```

1. Click on Open Console.

1. Click Administration > User Management

>     - Note: If you do not see `Manage Users`, you are probably using a [non-IAM instance](https://www.ibm.com/support/knowledgecenter/SS6NHC/com.ibm.swg.im.dashdb.security.doc/doc/iam.html). To confirm:

Reviewer: Can we remove this note now?  I'm not sure you can have a `Standard` or `Enterprise` plan that doesn't have IAM enabled

Reviewer: If we do keep it, I guess we should change `Manage Users` to "this menu option" or some such

Reviewer: paul's changes addressed this concern

> 
>     - Note: If you do not see `Manage Users`, you are probably using a [non-IAM instance](https://www.ibm.com/support/knowledgecenter/SS6NHC/com.ibm.swg.im.dashdb.security.doc/doc/iam.html). To confirm:

Author: ```suggestion

    - To confirm the IAM instance:
```

The [Db2 driver](https://repo1.maven.org/maven2/com/ibm/db2/jcc/11.5.0.0/jcc-11.5.0.0.jar) is able to execute a connectivity test to check the configuration of the combo of API-key/Service-Id/Db2-User-Id.

> #### Test IAM Access

Reviewer: If we keep both Db2 and IAM options above, then I think we should give them in the same order here.
Db2 first, then IAM.

Author: Yep, I've flipped them

1. Replace the following values with your service details:
    - `/path/to/db2jcc4.jar` : replace with the path to your driver jar.
    - `<DB-HOSTNAME>`: the hostname of your Db2 service from the Service Credentials page
    - `<userid>`: The userid to acecss the db
    - `<your_password>`: The password to access the db with
    - `<TRUSTSTORE-PASSWORD>`: the password for your truststore

1. Run in your favorite terminal, and you should see no errors in the output. You should see output like:

    ``` bash
    [jcc][10516][13709]Test Connection Successful.
    ```

### **Deploy the IBM FHIR Server schema**

Now that you've created the database and credentials, use the `fhir-persistence-schema` utility to deploy the IBM FHIR Server schema:

1. download the `fhir-persistence-schema` cli jar from the corresponding project release: https://github.com/IBM/FHIR/releases

2. create a properties file named db2.properties with the Db2 Admin connection info from IBM Cloud; for example:

    ``` sh
    db.host=1-2-3-4.databases.appdomain.cloud
    db.port=50001

>     db.database=bluedb

Reviewer: did they change the service?  in the past it was ALWAYS bludb (no `e`)

Author: great catch, fixed

    user=myuser
    password=mypassword
    sslConnection=true
    ```

3. execute the following commands:

    ``` sh
    java -jar schema/fhir-persistence-schema-*-cli.jar \
      --prop-file db2.properties --schema-name FHIRDATA --create-schemas
    java -jar schema/fhir-persistence-schema-*-cli.jar \
      --prop-file db2.properties --schema-name FHIRDATA --update-schema
    java -jar schema/fhir-persistence-schema-*-cli.jar \
      --prop-file db2.properties --schema-name FHIRDATA --grant-to FHIRSERVER
    java -jar schema/fhir-persistence-schema-*-cli.jar \
      --prop-file db2.properties --schema-name FHIRDATA --allocate-tenant default
    ```

> Note: Save the tenantKey from the `allocate-tenant` step above; this be needed to configure the IBM FHIR Server datasource in the next step

Reviewer: ```suggestion
4. save the tenantKey from the `allocate-tenant` step above; this is needed to configure the IBM FHIR Server datasource in the next step
```

For more information on using the fhir-persistence-schema cli jar, see https://github.com/IBM/FHIR/tree/main/fhir-persistence-schema/README.md.

### **Configuring an IBM FHIR Server datasource**

> The IBM FHIR Server uses a the native Open Liberty datasources. To configure a FHIR tenant datasource for a tenantKey, use the following template in the fhir-server-config.json:

Reviewer 2: Typo: "uses a the native Open"

> The IBM FHIR Server uses a the native Open Liberty datasources. To configure a FHIR tenant datasource for a tenantKey, use the following template in the fhir-server-config.json:

Author: ```suggestion
The IBM FHIR Server uses the native Open Liberty datasources. To configure a FHIR tenant datasource for a tenantKey, use the following template in the fhir-server-config.json:
```

``` json
        "persistence": {
            "datasources": {
                "default": {

>                     "tenantKey": "myTenantKey",
>                     "hints" : {

Author: ```suggestion
                    "tenantKey": "myTenantKey",
                    "currentSchema": "${DB_SCHEMA}",
                    "hints" : {
```

                        "search.reopt": "ONCE"
                    },

>                     "type": "db2",

Reviewer: very important
```suggestion
                    "type": "db2"
```

Reviewer: actually, we can leave the comma and add currentSchema:

```suggestion
                    "type": "db2",
                    "currentSchema": "${DB_SCHEMA}"
```

                }
            }
        }
```

Since release 4.3.2 you can use the `search.reopt` query optimizer hint (shown above) to improve the performance of certain search queries involving multiple search parameters. This optimization is currently only available for Db2. Valid values are "ALWAYS" and "ONCE". See Db2 documentation for `REOPT` for more details.

To configure the datasource.xml for db2 create a datasource.xml for the configDropins folder.

> #### For IAM User

Reviewer: again we should present these in a consistent order

Author: done

            securityMechanism="15"
            sslConnection="true" />
    </dataSource>
```

#### For Db2 Auth user

Create a file as the following: 

``` xml
<server>
    <!-- ============================================================== -->
    <!-- TENANT: default; DSID: default; TYPE: read-write               -->
    <!-- ============================================================== -->
    <dataSource id="fhirDefaultDefault" jndiName="jdbc/fhir_default_default" type="javax.sql.XADataSource" statementCacheSize="200" syncQueryTimeoutWithTransactionTimeout="true" validationTimeout="30s" isolationLevel="TRANSACTION_READ_COMMITTED">
        <jdbcDriver javax.sql.XADataSource="com.ibm.db2.jcc.DB2XADataSource" libraryRef="sharedLibDb2"/>
        <properties.db2.jcc
            erverName="${DB_HOSTNAME}"
            currentSchema="${DB_SCHEMA}"
            databaseName="${DB_NAME}"
            driverType="4"
            portNumber="${DB_PORT}"
            sslConnection="true" 
            user="fhirserver"
            password="mypassword"
         />
        <connectionManager maxPoolSize="200" minPoolSize="40"/>
    </dataSource>
</server>
```

> These files need to be created for each tenant along with a `fhir-server-config.json` in the tenant and configDropins/default folders.

Reviewer: ```suggestion
Each tenant datastore must have a corresponding dataSource definition and these dataSources must either follow the default jndiName pattern (`fhir_[tenantid]_[dsid]`) or the name must be explicitly configured in the corresponding section of `fhir-server-config.json`.
```

| portNumber | from the credential object select  `port` |
| databaseName | from the credential object select  `db`, generally always `BLUDB` |
| apiKey | from the created user in the assigned to the `fhiruser` group. Reference Section **FHIRSERVER User and API Key** |

> | securityMechanism | If using IAM, `15` generally set to 15 to trigger the `apiKey` use with IBM Cloud |

Reviewer: ```suggestion
| securityMechanism | If using IAM, set it to `15` to trigger the use of IAM-based `apiKey` authentication|
```

> | pluginName | If using IAM, `IBMIAMauth` fixed for use with the IBM Cloud|

Reviewer: ```suggestion
| pluginName | If using IAM, set to `IBMIAMauth`|
```
